new resource `azurerm_machine_learning_inference_cluster`

gro1m opened the pull request:
Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/11252

Test results:
PASS: TestAccInferenceCluster_requiresImport (858.19s)
--- PASS: TestAccInferenceCluster_custom_ssl_complete (899.78s)
--- PASS: TestAccInferenceCluster_basic (905.25s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/machinelearning     910.925s

Notes for the reviewer:
1. Update did not work -> *Tests* commented out, but functionality left in. I do not know where the mistake lies, but I guess I cannot let the update in, when the tests fail. Hoping the reviewer can suggest a fix for the tests.
2. Microsoft SSL Certificates did not work, even after passing cert and key as was requested by the error message -> not included.
3. As the Inference Cluster attaches to a Kubernetes Service of Type machinelearningservices.AKS unfortunately not all properties could be retrieved (via containerservice.AKS), namely "description", "cluster purpose" and the (custom) SSL Configuration.

katbyte commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,498 @@
+
+	v := input[0].(map[string]interface{})
+
+	identityType := machinelearningservices.ResourceIdentityType(v["type"].(string))
+
+	identity := machinelearningservices.Identity{
+		Type: identityType,
```
we can merge these
```suggestion
	identity := machinelearningservices.Identity{
		Type: machinelearningservices.ResourceIdentityType(v["type"].(string)),
```

gro1m replied:
done in the second commit

katbyte commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource_test.go:
```diff
@@ -0,0 +1,379 @@
+}
+
+/*
+func TestAccInferenceCluster_basicUpdate(t *testing.T) {
```
how come this is commented out?

gro1m replied:
I have commented in, but that is the problem I mentioned in the Slack Channel: I do not know why the update does not work.

gro1m replied:
Update does not work at the moment - also discussed with @aristosvo (as such is resolved).

katbyte commented on azurerm/internal/services/machinelearning/ssl_config/key.pem:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
could we put these in a testdata directory?

gro1m replied:
done in the second commit.

katbyte commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,498 @@
+				Default:  false,
+			},
+
+			"ssl_certificate_custom": {
```
i think we can remove custom as its implied?
```suggestion
			"ssl_certificate": {
```

gro1m replied:
I was thinking about doing a `ssl_certificate_custom` and `ssl_certificate_microsoft` block, but the Microsoft SSL configuration does not seem to work (just specifying `leaf_domain_label` still requests cert and key files). In the second commit I just go for `ssl` here and I removed the `ssl_enabled` parameter as I think this can be implied if you specify non-empty certificates and keys.

katbyte replied:
ssl_certificate/ssl_certificate_microsoft would work

gro1m replied:
Does that mean you are not happy with:
```bash
ssl {
      cert = file("testdata/cert.pem")
 	 key = file("testdata/key.pem")
 	 cname = "www.contoso.com"
   }
```
In principle the schema could be - correct me if I am wrong - improved in future (if the Microsoft bug does not occur - I think it is a bug) as follows:
```bash
			"ssl": {
				Type:     schema.TypeSet,
				Optional: true,
				ForceNew: true,
				MaxItems: 1,
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"cert": {
							Type:     schema.TypeString,
							Optional: true,
							Default:  "",
                                                        ConflictsWith: {"leaf_domain_label", "overwrite_existing_domain"},
						},
						"key": {
							Type:     schema.TypeString,
							Optional: true,
							Default:  "",
                                                        ConflictsWith: {"leaf_domain_label", "overwrite_existing_domain"},
						},
						"cname": {
							Type:     schema.TypeString,
							Optional: true,
							Default:  "",
                                                        ConflictsWith: {"leaf_domain_label", "overwrite_existing_domain"},
						},
                                               	"leaf_domain_label": {
							Type:     schema.TypeString,
							Optional: true,
							Default:  "",
                                                        ConflictsWith: {"cert", "key", "cname"},
						},
                                               	"overwrite_existing_domain": {
							Type:     schema.TypeString,
							Optional: true,
							Default:  "",
                                                        ConflictsWith: {"cert", "key", "cname"},
						},
					},
				},
			},
```
What do you think is best for possible future enhancements: 2 separate blocks with now only an "ssl_certificate" block or everything in the "ssl" block or if you want "ssl_certificate" block?

katbyte commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,498 @@
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"cert": {
```
is this the certificate data?
```suggestion
						"data": {
```
(as certificate is in the block name)

gro1m replied:
In the second commit I changed the block name to `ssl`, as if there is an error, it is going to be of the form `ssl.cert` or `ssl.key` is missing.

katbyte reviewed (changes requested):
Thanks @gro1m - i've given this a quick review and it's off the a great start!

gro1m commented:
I included the update now, but I expect it to fail:

TF_ACC=1 go test -v ./azurerm/internal/services/machinelearning -run=TestAccInferenceCluster -timeout 60m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/05/05 08:24:20 [DEBUG] not using binary driver name, it's no longer needed
2021/05/05 08:24:20 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccInferenceCluster_basic
=== PAUSE TestAccInferenceCluster_basic
=== RUN   TestAccInferenceCluster_requiresImport
=== PAUSE TestAccInferenceCluster_requiresImport
=== RUN   TestAccInferenceCluster_complete
=== PAUSE TestAccInferenceCluster_complete
=== RUN   TestAccInferenceCluster_basicUpdate
=== PAUSE TestAccInferenceCluster_basicUpdate
=== RUN   TestAccInferenceCluster_completeUpdate
=== PAUSE TestAccInferenceCluster_completeUpdate
=== CONT  TestAccInferenceCluster_basicUpdate
=== CONT  TestAccInferenceCluster_completeUpdate
=== CONT  TestAccInferenceCluster_complete
=== CONT  TestAccInferenceCluster_requiresImport
=== CONT  TestAccInferenceCluster_basic
Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799178" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797524" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420791885" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797028" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799508" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799178" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797524" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420791885" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797028" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799508" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799178" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799508" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797028" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420791885" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797524" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799178" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799178" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797028" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797028" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420791885" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420791885" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797524" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797524" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799508" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799178" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797524" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420791885" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797028" "Microsoft.MachineLearningServices" map[]}=== CONT  TestAccInferenceCluster_basicUpdate
    testing.go:620: Step 3/4 error: Error running apply: 
        Error: A resource with the ID "/subscriptions/b609167b-ec1b-406c-accc-dc6cb5df9f43/resourceGroups/acctestRG-ml-210505082420799178/providers/Microsoft.MachineLearningServices/workspaces/acctest-MLW2105050824207991/computes/AIC-21050578" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_machine_learning_inference_cluster" for more information.
        
          on ../../../../../../../../var/folders/t2/y0zhp0x97q5bk_s_sjq4b3zh0000gn/T/tftest210209280/work723527519/config879850977/terraform_plugin_test.tf line 111, in resource "azurerm_machine_learning_inference_cluster" "test":
         111: resource "azurerm_machine_learning_inference_cluster" "test" {
        
        
Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799508" "Microsoft.MachineLearningServices" map[]}=== CONT  TestAccInferenceCluster_completeUpdate
    testing.go:620: Step 3/4 error: Error running apply: 
        Error: A resource with the ID "/subscriptions/b609167b-ec1b-406c-accc-dc6cb5df9f43/resourceGroups/acctestRG-ml-210505082420797028/providers/Microsoft.MachineLearningServices/workspaces/acctest-MLW2105050824207970/computes/AIC-21050528" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_machine_learning_inference_cluster" for more information.
        
          on ../../../../../../../../var/folders/t2/y0zhp0x97q5bk_s_sjq4b3zh0000gn/T/tftest210209280/work965243913/config501865756/terraform_plugin_test.tf line 111, in resource "azurerm_machine_learning_inference_cluster" "test":
         111: resource "azurerm_machine_learning_inference_cluster" "test" {
        
        
Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420799178" "Microsoft.MachineLearningServices" map[]}Debug: id = &{"b609167b-ec1b-406c-accc-dc6cb5df9f43" "acctestRG-ml-210505082420797028" "Microsoft.MachineLearningServices" map[]}--- PASS: TestAccInferenceCluster_requiresImport (947.45s)
--- PASS: TestAccInferenceCluster_basic (1009.88s)
--- PASS: TestAccInferenceCluster_complete (1015.03s)
--- FAIL: TestAccInferenceCluster_basicUpdate (1034.19s)
--- FAIL: TestAccInferenceCluster_completeUpdate (1038.56s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/machinelearning     1043.327s
FAIL
make: *** [acctests] Error 1

I have no clue what I am doing wrong, so if someone can help me with this I would be very happy

gro1m commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,498 @@
+				ForceNew: true,
+			},
+
+			"kubernetes_cluster_rg": {
```
Done in the second commit, I did it as well for the Machine Learning Workspace ID.

gro1m commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,498 @@
+	machine_learning_workspaces_client := meta.(*clients.Client).MachineLearning.WorkspacesClient
+	machine_learning_compute_client := meta.(*clients.Client).MachineLearning.MachineLearningComputeClient
+	kubernetes_clusters_client := meta.(*clients.Client).Containers.KubernetesClustersClient
+	node_pools_client := meta.(*clients.Client).Containers.AgentPoolsClient
```
Done in the second commit.

gro1m commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,498 @@
+	if err := future.WaitForCompletionRef(ctx, machine_learning_compute_client.Client); err != nil {
+		return fmt.Errorf("error waiting for creation of Inference Cluster %q in workspace %q (Resource Group %q): %+v", name, workspace_name, resource_group_name, err)
+	}
+	resp, ml_get_err := machine_learning_compute_client.Get(ctx, resource_group_name, workspace_name, name)
```
yes, I do not remember what the problem was...

gro1m commented:
@katbyte I thought the git diff was incorrect (i.e. all lines changed), but I was looking at commit 1 instead of commit 2. So then I pushed 3 commits to fix that just to realize that there were unneccessary. In order not to pollute this PR I reset hard these three commits with:
```bash
git reset --hard 94639c6f1
git push --force-with-lease
```
The only problem we have now is that it states "16 workflows awaiting approval" instead of "8 workflows awaiting approval" - maybe has to do something with my "hard reset" - sorry about that.
 Apart from that, I should have addressed all your comments and I would be happy if you could further review:)

gro1m commented on azurerm/internal/services/machinelearning/parse/inference_cluster.go:
```diff
@@ -7,7 +7,7 @@ import (
 )
```
How can I (possibly) resolve this golint error on line 5? - All other linting-related errors should have been addressed

katbyte commented:
THanks @gro1m - have 5 failing tests now:
![image](https://user-images.githubusercontent.com/1638467/117510364-cdc5eb80-af40-11eb-8a98-bb2998a3b7ea.png)

aristosvo commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,502 @@
+	name := d.Get("name").(string)
+
+	// Get Machine Learning Workspace Name and Resource Group from ID
+	ml_workspace_id := d.Get("machine_learning_workspace_id").(string)
+	ws_id, err := parse.WorkspaceID(ml_workspace_id)
+	if err != nil {
+		return err
+	}
+
+	workspace_name := ws_id.Name
+	resource_group_name := ws_id.ResourceGroup
```
Variables should be camelCase, I'd use workspaceID directly instead of initiating the variables here already
```suggestion
	workspaceIDUnparsed := d.Get("machine_learning_workspace_id").(string)
	workspaceID, err := parse.WorkspaceID(workspaceIDUnparsed)
	if err != nil {
		return err
	}
```

aristosvo commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,502 @@
+	resource_group_name := ws_id.ResourceGroup
+
+	// Check if Inference Cluster already exists
+	existing, err := mlComputeClient.Get(ctx, resource_group_name, workspace_name, name)
```
Use of workspaceID instead of variables
```suggestion
	existing, err := mlComputeClient.Get(ctx, workspaceID.ResourceGroup, workspaceID.name, name)
```

aristosvo commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,502 @@
+	existing, err := mlComputeClient.Get(ctx, resource_group_name, workspace_name, name)
+	if err != nil {
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return fmt.Errorf("error checking for existing Inference Cluster %q in Workspace %q (Resource Group %q): %s", name, workspace_name, resource_group_name, err)
```
Same
```suggestion
			return fmt.Errorf("error checking for existing Inference Cluster %q in Workspace %q (Resource Group %q): %s", name, workspaceID.name, workspaceID.ResourceGroup, err)
```

aristosvo commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,502 @@
+	}
+
+	// Get SKU from Workspace
+	aml_ws, err := mlWorkspacesClient.Get(ctx, resource_group_name, workspace_name)
```
Same
```suggestion
	respWorkspace, err := mlWorkspacesClient.Get(ctx, workspaceID.ResourceGroup, workspaceID.Name)
```

aristosvo commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,502 @@
+	if err != nil {
+		return err
+	}
+	sku := aml_ws.Sku
```
Same
```suggestion
	sku := respWorkspace.Sku
```

aristosvo commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,502 @@
+	sku := aml_ws.Sku
+
+	// Get Kubernetes Cluster Name and Resource Group from ID
+	aks_id := d.Get("kubernetes_cluster_id").(string)
+	aks_id_details, err := parse.KubernetesClusterID(aks_id)
+	if err != nil {
+		return err
+	}
+	kubernetes_cluster_name := aks_id_details.ManagedClusterName
+	kubernetes_cluster_rg := aks_id_details.ResourceGroup
```
Same here
```suggestion
	aksIDUnparsed := d.Get("kubernetes_cluster_id").(string)
	aksID, err := parse.KubernetesClusterID(aksIDUnparsed)
	if err != nil {
		return err
	}
```

aristosvo reviewed (changes requested):
A few comments to give you an idea how I'd start here to rework it. This makes it easier to cut to functionality later together

aristosvo commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,495 @@
+		return err
+	}
+
+	nodePoolList, _ := poolsClient.List(ctx, id.ResourceGroup, aksId.ManagedClusterName)
+	unparsedPoolId := *(nodePoolList.Values()[0].ID)
+	poolId, err := parse.NodePoolID(unparsedPoolId)
+	if err != nil {
+		return err
+	}
```
I think we should remove the node pool (and the configuration based on the node pool) from the configuration. This is not used when attaching a cluster from the [commandline](https://docs.microsoft.com/en-us/cli/azure/ml/computetarget/attach?view=azure-cli-latest), so I expect the configuration not necessary to the functioning of the inference cluster. WDYT?

For a reference I use both the `attach` and [`update` docs](https://docs.microsoft.com/en-us/cli/azure/ml/computetarget/update?view=azure-cli-latest#az_ml_computetarget_update_aks) for an AKS compute target, so maybe we should expose some other knobs to effectively update the configuration.

gro1m replied:
Hmm... Yeah I am not sure how to determine the subnet id from the existing AKS service in another way. The other thing for what I used it is to determine agent count and vm size, which I thought you could infer from the Azure Kubernetes node pool, but maybe you should do that differently. You might even need to use them as input parameters (if not inferred from a node pool), which would leave us with the problem of how to determine the subnet id from an existing AKS.

aristosvo reviewed (changes requested):
Round 2 😄 

Your changes make it a lot easier for me to go through your code, just some small remarks on minor details left.

I went a bit further and inspected your implementation against the API from a TF point of view. What I see is that you implemented and filled almost all possible fields of the objects. I have a suspicion it might be too much for a first implementation (and probably not necessary), especially as it requires information as input which is not exposed from the resource itself.

Nice job anyway, I think it's almost there if we can tweak the input parameters a bit!

gro1m commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,495 @@
+		"Test": "DevTest",
+		"Prod": "FastProd",
+	}
+	clusterPurpose := machinelearningservices.ClusterPurpose(mapClusterPurpose[unmappedclusterPurpose])
```
done

gro1m commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,495 @@
+	}
+
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+	id := parse.NewInferenceClusterID(subscriptionId, workspaceID.ResourceGroup, workspaceID.Name, name)
```
done

gro1m commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource.go:
```diff
@@ -0,0 +1,495 @@
+	}
+
+	unparsedAksId := *(aksResp.Values()[0].ID)
+
```
Ah that does it :) Ok yes, done as well.

aristosvo commented:
```
❯ make build && make acctests SERVICE='machinelearning' TESTARGS='-run=TestAccInferenceCluster_complete'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
go generate ./azurerm/internal/services/...
go generate ./azurerm/internal/provider/
go install
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/machinelearning -run=TestAccInferenceCluster_complete -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/05/14 15:28:15 [DEBUG] not using binary driver name, it's no longer needed
2021/05/14 15:28:16 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccInferenceCluster_complete
=== PAUSE TestAccInferenceCluster_complete
=== CONT  TestAccInferenceCluster_complete
--- PASS: TestAccInferenceCluster_complete (917.72s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/machinelearning     921.445s
```

gro1m commented:
> THanks @gro1m - have 5 failing tests now:
> ![image](https://user-images.githubusercontent.com/1638467/117510364-cdc5eb80-af40-11eb-8a98-bb2998a3b7ea.png)

@katbyte With the kind help of @aristosvo (we did some programming together today to put this into a better state), things should work now and I would be happy if could go through this PR again.

I have the following remarks, that are either questions from my side or implementation points that might be of interest to you:
1. We could not do any update, but also the az cli does not provide sensible updates really, and as this is an attached resource of AKS we think we should be ok.
2. In principle, the portal offers two variants of configuring an inference cluster: a) attach to AKS b) pass configuration that also creates an AKS. For simplicity reasons, we decided to go with option a).
3. We cannot read the ssl configuration, as the cert and key parameter of the ssl list are redacted. We tried to set `ssl.0.cname` but it was not possible, as there is an error by the test, that you can only set the full list.
4. We think that it should work now, strangely there are 8 failing and 8 successful workflows at the same time.
5. Would look forward to your suggestions regarding README. 
- I am a bit unsure, if it is sensible just to pass a long ID or if I should put resource configurations for Machine Learning Workspace and Kubernetes Cluster, the former necessitating then also defining application insights, storage accounts and key vault, which might clutter the whole stuff.
- How to best note that this resource cannot be updated
6. There is no test for production, as this could be a bit expensive. Also there should be at least 12 = agent_count * vm_cores for "FastProd" cluster purpose. There will be just an error creating the compute and the portal will give a reply. I hope that is fine, because we did not find a way to really check that reasonably.

aristosvo commented:
@katbyte @gro1m I took the liberty to update the docs and tests based on the last 2 comments to make it easier to comment on/review it. Hope you don't mind 🥺

gro1m commented on azurerm/internal/services/machinelearning/machine_learning_inference_cluster_resource_test.go:
```diff
@@ -138,10 +174,17 @@ resource "azurerm_machine_learning_inference_cluster" "import" {
+}
+
+func (r InferenceClusterResource) templateFastProd(data acceptance.TestData) string {
+	return r.template(data, "SStandard_D3_v2", 3)
```
@aristosvo is this a spelling mistake "SStandard_D3_v2". Should it not be: "Standard_D3_v2" - I guess you did not run this test:)

aristosvo replied:
Got me 😅 Running now!

gro1m commented:
Did also test here the latest stand and every test passes.
```bash
TF_ACC=1 go test -v ./azurerm/internal/services/machinelearning -run=TestAccInferenceCluster -timeout 60m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/05/19 21:43:59 [DEBUG] not using binary driver name, it's no longer needed
2021/05/19 21:43:59 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccInferenceCluster_basic
=== PAUSE TestAccInferenceCluster_basic
=== RUN   TestAccInferenceCluster_requiresImport
=== PAUSE TestAccInferenceCluster_requiresImport
=== RUN   TestAccInferenceCluster_complete
=== PAUSE TestAccInferenceCluster_complete
=== RUN   TestAccInferenceCluster_completeProduction
=== PAUSE TestAccInferenceCluster_completeProduction
=== CONT  TestAccInferenceCluster_basic
=== CONT  TestAccInferenceCluster_completeProduction
=== CONT  TestAccInferenceCluster_complete
=== CONT  TestAccInferenceCluster_requiresImport
--- PASS: TestAccInferenceCluster_complete (898.54s)
--- PASS: TestAccInferenceCluster_requiresImport (917.88s)
--- PASS: TestAccInferenceCluster_completeProduction (951.42s)
--- PASS: TestAccInferenceCluster_basic (952.04s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/machinelearning     957.246s
```

gro1m commented:
@ArcturusZhang Am I right that there is no sense to update anything here?
I checked with @aristosvo, but we did not find anything useful to update nor did it work.
The only parameters that we found that maybe could be updatable were:
1. `SSL certificates` -> we think that this should force the creation of a new inference cluster.
2. `tags` -> also do not seem to be updatable nor setable via the Portal UI.

katbyte reviewed (approved):
Thanks @gro1m - LGTM 👍
